### PR TITLE
Fire - Roshni 

### PR DESCRIPTION
**Pull request by roshni-patel**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? | 

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | I wrote the self.top_ten method in the work model. It returns the top ten works from a given category based on the amount of votes.
Describe how you approached testing that model method. What edge cases did you come up with? | I first started by writing nominal test cases, e.g. checking that ten works could be returned and that they all belonged to the same category. I then moved on to edge cases such as ties in votes, no votes overall, and less than ten works available. 
What are `session` and `flash`? What is the difference between them? | Session and flash are both hash-like objects involved in the storage of data for multiple response-request cycles or one response-request cycle, respectively. 
What was one thing that you gained more clarity on through this assignment? | I gained much more clarity on writing model and controller tests as well as how the test database works.
What is the Heroku URL of your deployed application? | https://roshni-media-ranker.herokuapp.com/

### CODE REVIEW

File: app/models/vote.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class Vote < ApplicationRecord
+  belongs_to :work
+  belongs_to :user
+
+  validates :user_id, uniqueness: { scope: :work_id, message: "user: has already voted for this work" }
```
Comment (kaidamasaki) [inline app/models/vote.rb:5]: Nice use of a scope.  😃

File: app/views/works/edit.html.erb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+<h2>Edit This <%= @work.category.capitalize %></h2>
```
Comment (kaidamasaki) [inline app/views/works/edit.html.erb:1]: You can capitalize things in CSS: https://developer.mozilla.org/en-US/docs/Web/CSS/text-transform

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
+class UsersController < ApplicationController
+  def index
+    if session[:user_id]
+      @users = User.all
+    else
+      flash.now[:error] = "You must be logged in to see this page"
```
Comment (kaidamasaki) [inline app/controllers/users_controller.rb:6]: Usually you would redirect but this is an interesting solution that seems like it will work just as well.  

Good job.  😃

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+class Work < ApplicationRecord
+  has_many :votes, dependent: :destroy
+  has_many :users, through: :votes
+
+  validates :title, presence: true, uniqueness: { scope: :category }
+  validates :category, presence: true
+
+  def self.spotlight
+    return Work.all.max_by { |work| work.votes.count}
```
Comment (kaidamasaki) [inline app/models/work.rb:9]: Nice use of `max_by`!

File: app/models/work.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  def self.sorted(category)
+    return Work.where(category: category).sort_by { |work| work.votes.count }.reverse
+  end
+
+  def vote_date
+    vote = Vote.find_by(work_id: self.id)
+    return vote.created_at.strftime("%b %d, %Y")
+  end
```
Comment (kaidamasaki) [inline app/models/work.rb:23]: Since this is formatting a string it should have probably lived in a view helper.

File: test/models/user_test.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  describe "relationships" do
+    before do
+      new_user.save
+      @work_1 = Work.create(category: "book", title: "test title", creator: "test creator", publication_year: 2020, description: "test description")
+      @work_2 = Work.create(category: "album", title: "test title", creator: "test creator", publication_year: 2020, description: "test description")
+
+      Vote.create(user_id: new_user.id, work_id: @work_1.id)
+      Vote.create(user_id: new_user.id, work_id: @work_2.id)
```
Comment (kaidamasaki) [inline test/models/user_test.rb:39]: I recommend using `create!` in your `before` blocks so that you get an exception if it is invalid instead of a silent failure.

**Review by kaidamasaki — COMMENTED** [review]
Excellent job!  I don't have too much to say about your project, things were clean and clear and you solved things in a number of elegant ways.

# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️?
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | Missing fixtures.
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
